EIPS/eip-9999.md

```diff
@@ -0,0 +1,92 @@
+status: Draft
+type: Standards Track
+category: Core
+created: 2025-03-31
+---
+
+## Abstract
+
+This EIP proposes adding block-level access lists. By including a complete list of all addresses and storage keys accessed during a block, along with their initial values, in the block header, we can improve execution efficiency and enable better parallelization.
+
+## Motivation
+
+Currently, transactions without an explicit transaction access list cannot be efficiently parallelized, as the execution engine cannot determine in advance which addresses and storage slots will be accessed. While transaction-level access lists exist, they are not enforced, making it difficult to optimize execution pipelines.
+
+By enforcing access lists at the block level, we shift the responsibility of declaring storage accesses to the block builder. This enables the execution engine to efficiently parallelize transaction execution, knowing in advance the exact scope of storage interactions for each transaction. 
+
+With access lists at the block level, transactions can be scheduled and executed in parallel, allowing for performance improvements.
+
+## Specification
+
+### Block Header Modification
+
+Introduce a new field in the block header:
+
+```python
+class BlockBody:
+    ...
+    # Each access → ordered list of tx indices and associated values
+    block_level_accesses: Dict[
+        Tuple[Address, StorageKey],
+        Tuple[
+            List[int],  # The transaction indices that accessed this slot
+            List[Union[Value, Tuple[Value, Optional[Value]]]]  # The corresponding read or write values
+        ]
+    ]
+```
+
+The block_level_access_list maps each accessed address and storage key to the index of the transaction that accesses it, along with the corresponding storage value. This structure enables efficient validation and execution by explicitly tying storage interactions and their values to specific transactions.
+
+### State Transition Function
+
+Modify the state transition function to validate the block-level access lists asynchronously:
+
+```python 
+def state_transition(block: Block) -> None:
+    computed_access_list = {}
+
+    for idx, tx in enumerate(block.transactions):
+        accessed_items = execute_transaction(tx)  # returns {(address, slot): (value_before, optional_value_after)}
+        for key, val in accessed_items.items():
+            computed_access_list[key] = (idx, val)
+
+    if computed_access_list != block.header.block_level_access_list:
+        raise InvalidBlock("Mismatch in block-level access list or values.")
+
+```
+
+Transactions must not read or write to any address or storage key not declared in the block-level access list. Additionally, the block is invalid if the access list contains entries for addresses or storage keys that were not actually accessed, or if the recorded values deviate from those observed during execution.
+
+## Rationale
+
```

Comment:
Since we now have an access list for address+storage at the block level, can we now drop EIP-2929 logic entirely? So treat all accounts+storage as "warm"?

Reply:
I'm not entirely sure about that because if we treat everything as warm, we also allow many more reads and writes which again affects the size of the access list. Maybe we should treat everything cold and focus on moving towards a better, block-level pricing mechanism for storage.